File: tests/r-process/r-process.cpp
/// \file r-process.cpp
/// \author jlippuner
/// \since Jul 1, 2014
///
/// \brief
///
///

#include <cstdio>
#include <chrono>
#include <cmath>
#include <fstream>
#include <map>
#include <thread>
#include <vector>

#include "BuildInfo.hpp"
#include "EquationsOfState/HelmholtzEOS.hpp"
#include "EquationsOfState/SkyNetScreening.hpp"
#include "Network/NSE.hpp"
#include "Network/ReactionNetwork.hpp"
#include "Network/TemperatureDensityHistory.hpp"
#include "Utilities/FlushStdout.hpp"
#include "Reactions/REACLIBReactionLibrary.hpp"
#include "Utilities/Profiler.hpp"
#include "Utilities/FloatingPointExceptions.hpp"

int main(int, char**) {
  FloatingPointExceptions::Enable();

  auto nuclib = NuclideLibrary::CreateFromWebnucleoXML(
      SkyNetRoot + "/data/webnucleo_nuc_v2.0.xml");

  NetworkOptions opts;
  opts.MaxYChangePerStep = 0.1;
  opts.SmallestYUsedForDtCalculation = 1.0E-4;
  opts.ConvergenceCriterion = NetworkConvergenceCriterion::Mass;
  opts.SmallestYUsedForErrorCalculation = 1.0E-20;
  opts.MaxDtChangeMultiplier = 2.0;
  opts.MinDt = 1.0E-16;
  opts.IsSelfHeating = true;

  REACLIBReactionLibrary weakReactionLibrary(SkyNetRoot + "/data/reaclib",
      ReactionType::Weak, false, LeptonMode::TreatAllAsDecayExceptLabelEC,
      "Weak reactions", nuclib, opts, true);
  REACLIBReactionLibrary strongReactionLibrary(SkyNetRoot + "/data/reaclib",
      ReactionType::Strong, true, LeptonMode::TreatAllAsDecayExceptLabelEC,
      "Strong reactions", nuclib, opts, true);
  REACLIBReactionLibrary symmetricFission(
      SkyNetRoot + "/data/netsu_panov_symmetric_0neut",
      ReactionType::Strong, false, LeptonMode::TreatAllAsDecayExceptLabelEC,
      "Symmetric neutron induced fission with 0 free neutrons", nuclib, opts,
      false);
  REACLIBReactionLibrary spontaneousFission(
      SkyNetRoot + "/data/netsu_sfis_Roberts2010rates",
      ReactionType::Strong, false, LeptonMode::TreatAllAsDecayExceptLabelEC,
      "Spontaneous fission", nuclib, opts, false);

  auto hist = TemperatureDensityHistory::CreateFromFile("temp_rho_vs_time");

  HelmholtzEOS helmEOS(SkyNetRoot + "/data/helm_table.dat");
  SkyNetScreening screen(nuclib);

  ReactionNetwork net(nuclib, { &weakReactionLibrary, &strongReactionLibrary,
      &symmetricFission, &spontaneousFission }, &helmEOS, &screen, opts);

  //auto output = net.EvolveFromNSE(hist, "SkyNet_output");
  auto output = net.EvolveFromNSE(hist.StartTime(), hist.StartTime() + 1.E-10,
      &hist.TemperatureVsTime(), &hist.DensityVsTime(), hist.Ye(),
      "SkyNet_output");

  std::vector<double> finalY = output.FinalY();

//  FILE * fout = fopen("fin", "w");
//  for (unsigned int i = 0; i < finalY.size(); ++i)
//    fprintf(fout, "%.20E\n", finalY[i]);
//  fclose(fout);

  std::vector<double> expectedY(finalY.size());
  {
    std::ifstream ifs("final_abundances", std::ifstream::in);
    for (unsigned int i = 0; i < expectedY.size(); ++i)
      ifs >> expectedY[i];

    ifs.close();
  }

  double maxError = 0.0;
  for (unsigned int i = 0; i < finalY.size(); ++i) {
    if (std::isnan(finalY[i])) {
      printf("got NaN\n");
      return EXIT_FAILURE;
    }
    if (finalY[i] < opts.SmallestYUsedForErrorCalculation)
      continue;
    double error = fabs(finalY[i] - expectedY[i])
        / expectedY[i];
    maxError = std::max(maxError, error);
  }

  printf("max fractional error = %.10E\n", maxError);

  if (maxError < 1.0E-8)
    return EXIT_SUCCESS;
  else
    return EXIT_FAILURE;
}

File: examples/r-process.py
#!/usr/bin/env python

from SkyNet import *
import numpy as np

nuclib = NuclideLibrary.CreateFromWebnucleoXML(SkyNetRoot
    + "/data/webnucleo_nuc_v2.0.xml")

opts = NetworkOptions()
opts.ConvergenceCriterion = NetworkConvergenceCriterion.Mass
opts.MassDeviationThreshold = 1.0E-10
opts.IsSelfHeating = True
opts.EnableScreening = True

screen = SkyNetScreening(nuclib)
helm = HelmholtzEOS(SkyNetRoot + "/data/helm_table.dat")

strongReactionLibrary = REACLIBReactionLibrary(SkyNetRoot + "/data/reaclib",
  ReactionType.Strong, True, LeptonMode.TreatAllAsDecayExceptLabelEC,
  "Strong reactions", nuclib, opts, True)
symmetricFission = REACLIBReactionLibrary(SkyNetRoot
  + "/data/netsu_panov_symmetric_0neut", ReactionType.Strong, False,
  LeptonMode.TreatAllAsDecayExceptLabelEC,
  "Symmetric neutron induced fission with 0 neutrons emitted", nuclib, opts,
  False)
spontaneousFission = REACLIBReactionLibrary(SkyNetRoot +
  "/data/netsu_sfis_Roberts2010rates", ReactionType.Strong, False,
  LeptonMode.TreatAllAsDecayExceptLabelEC, "Spontaneous fission", nuclib, opts,
  False)

# use only REACLIB weak rates
weakReactionLibrary = REACLIBReactionLibrary(SkyNetRoot + "/data/reaclib",
    ReactionType.Weak, False, LeptonMode.TreatAllAsDecayExceptLabelEC,
    "Weak reactions", nuclib, opts, True)

# or use the following code to use FFN rates and weak REACLIB rates pre-computed
# with the <SkyNetRoot>/examples/precompute_reaction_libs.py script

# ffnMesaReactionLibrary = FFNReactionLibrary.ReadFromDisk(
#     "ffnMesa_with_neutrino", opts)
# ffnReactionLibrary = FFNReactionLibrary.ReadFromDisk(
#     "ffn_with_neutrino_ffnMesa", opts)
# weakReactionLibrary = REACLIBReactionLibrary.ReadFromDisk(
#     "weak_REACLIB_with_neutrino_ffnMesa_ffn", opts)

# add neutrino reactions, if desired (will need a neutrino distribution
# function set with net.LoadNeutrinoHistory(...))
# neutrinoLibrary = NeutrinoReactionLibrary(SkyNetRoot
#     + "/data/neutrino_reactions.dat", "Neutrino interactions", nuclib, opts,
#     1.e-2, False, True)

reactionLibraries = [strongReactionLibrary, symmetricFission,
    spontaneousFission, weakReactionLibrary]

# reactionLibraries = [strongReactionLibrary, symmetricFission,
#     spontaneousFission, weakReactionLibrary, neutrinoLibrary,
#     ffnMesaReactionLibrary, ffnReactionLibrary]

net = ReactionNetwork(nuclib, [weakReactionLibrary, strongReactionLibrary,
    symmetricFission, spontaneousFission], helm, screen, opts)

T0 = 6.0    # initial temperature in GK
Ye = 0.01   # initial Ye
s = 10.0    # initial entropy in k_B / baryon
tau = 7.1   # expansion timescale in ms

# run NSE with the temperature and entropy to find the initial density
nse = NSE(net.GetNuclideLibrary(), helm, screen)
nseResult = nse.CalcFromTemperatureAndEntropy(T0, s, Ye)

densityProfile = ExpTMinus3(nseResult.Rho(), tau / 1000.0);

output = net.EvolveSelfHeatingWithInitialTemperature(nseResult.Y(), 0.0, 1.0E9,
    T0, densityProfile, "SkyNet_r-process_python")

NetworkOutput.MakeDatFile("SkyNet_r-process_python.h5")

YvsA = np.array(output.FinalYVsA())
A = np.arange(len(YvsA))

np.savetxt("final_y_r-process_python", np.array([A, YvsA]).transpose(),
    "%6i  %30.20E")
